### PR TITLE
Virtual baton

### PR DESCRIPTION
**Pull request by howard-stearns**
Updates tribbles example to arrange for only one of the participating clients to be the one to update the colors. This is done through library/virtualBaton.js, which can be claimed by any participant, but only held by one at a time (even as participants and servers come and go).

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: b08f3644474af7185650c33a5fd61b244fd89109 available at: https://gist.github.com/db791c2c7496622bf07a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.dmg

**Comment by howard-stearns** [comment]
There's also a mocha test suite at tests/mocha. (It's not possible to develop this kind of software without tests.)

https://app.asana.com/0/32622044445063/85114103526644

See also https://github.com/highfidelity/hifi/pull/6903

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4a50d7c456c441936ea55820d3db3baf76c7b737 available at: https://gist.github.com/db791c2c7496622bf07a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 61e521bc63b2ebfc4ad24bcf328e1cbf223d21bc available at: https://gist.github.com/db791c2c7496622bf07a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.dmg

**Comment by howard-stearns** [comment]
DNM b/c I'm doing some testing....

**Comment by hifi-gustavo** [comment]
Build console output for commit: 9670cf3fe32baf9da0e992deafc48a98aa94b2af available at: https://gist.github.com/db791c2c7496622bf07a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d174e863472e3241dce57a092b1ef25840846a77 available at: https://gist.github.com/db791c2c7496622bf07a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.dmg

**Comment by howard-stearns** [comment]
"should" be ok now. I had a brain-o, and a funky server.

**Comment by hifi-gustavo** [comment]
Build console output for commit: cecf2fb611c474df895d413ab5a7505200c419f4 available at: https://gist.github.com/db791c2c7496622bf07a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.dmg

**Comment by howard-stearns** [comment]
For testing, instead of mucking about with coordinating build and file system between multiple machines, one might:
1. Copy the tribbles script (that's plural) from https://raw.githubusercontent.com/howard-stearns/hifi/virtualBaton/examples/tests/performance/tribbles.js
2. Change the script reference near the bottom to the following so that its available on multiple machines: https://raw.githubusercontent.com/howard-stearns/hifi/virtualBaton/examples/entityScripts/tribble.js (That's tribble singular.)
3. Run that tribbles as an interface script from one machine, while observing in zero or more other machines. All the machines should see all the balls change color slowly, even when any of the machines leave and/or enter. (The balls stay present for two minutes. You can change that at the top of the tribbles script that you downloaded.)

The count of 100 tribble balls is probably at the limit of what we can handle -- it may be over the limit on your machine/network. You can try smaller numbers by changing the parameter at the top of the tribbles script that you downloaded.

File: examples/entityScripts/tribble.js [hunk 1]
```diff
@@ -35,21 +37,43 @@
...
         if (!shutdown) { Script.setTimeout(move, nextChange); }
     }
+    function startUpdate() {
+        print('startUpdate', entityID);
+        hasUpdate = true;
+        Script.update.connect(update);
+    }
+    function stopUpdate() {
+        print('stopUpdate', entityID, hasUpdate);
+        if (!hasUpdate) { return; }
```
Comment (ZappoMan) [inline examples/entityScripts/tribble.js]: coding standard: braces on separate lines.

File: examples/entityScripts/tribble.js [hunk 2]
```diff
@@ -35,21 +37,43 @@
         Entities.editEntity(entityID, newData);
         if (!shutdown) { Script.setTimeout(move, nextChange); }
```
Comment (ZappoMan) [inline examples/entityScripts/tribble.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+//  virtualBaton(), claim(), release().
+//
+
+// Answers a new virtualBaton for the given parameters, of which 'key'
+// is required.
+function virtualBatonf(options) {
+    // Answer averages (number +/- variability). Avoids having everyone act in lockstep.
+    function randomize(number, variability) {
+        var allowedDeviation = number * variability;
+        return number - allowedDeviation + (Math.random() * 2 * allowedDeviation);
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: magic number... why 2 here?

File: examples/libraries/virtualBaton.js [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+        recheckInterval = options.recheckInterval || randomize(500, 0.2), // ms. Check that winners remain connected.
+        // If you supply your own instanceId, you might also supply a connectionTest that answers
+        // truthy iff the given id is still valid and connected, and is run at recheckInterval. You
+        // can use exports.validId (see below), and the default answers truthy if id is valid or a
+        // concatenation of two valid ids. (This handles the most common cases of instanceId being
+        // either (the default) MyAvatar.sessionUUID, an entityID, or the concatenation (in either
+        // order) of both.)
+        connectionTest = options.connectionTest || function connectionTest(id) {
+            var idLength = 38;
+            if (id.length === idLength) { return exports.validId(id); }
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+            return (id.length === 2 * idLength) && exports.validId(id.slice(0, idLength)) && exports.validId(id.slice(idLength));
+        };
+
+    if (!batonName) {
+        throw new Error("A virtualBaton must specify a batonName.");
+    }
+    // Truthy if id exists as either a connected avatar or valid entity.
+    exports.validId = function validId(id) {
+        var avatar = avatarList.getAvatar(id);
+        if (avatar && (avatar.sessionUUID === id)) { return true; }
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+        var properties = entities.getEntityProperties(id, ['type']);
+        return properties && properties.type;
+    };
+
+    // Various logging, controllable through options.
+    function debug() { // Display the arguments not just [Object object].
+        log.apply(null, [].map.call(arguments, JSON.stringify));
+    }
+    function debugFlow() {
+        if (options.debugFlow) { debug.apply(null, arguments); }
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+    }
+    function debugSend(destination, operation, data) {
+        if (options.debugSend) { debug('baton:', batonName, instanceId, 's=>', destination, operation, data); }
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+    }
+    function debugReceive(senderID, operation, data) { // senderID is client sessionUUID -- not necessarily instanceID!
+        if (options.debugReceive) { debug('baton:', batonName, senderID, '=>r', instanceId, operation, data); }
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 9]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    // Messages: Just synactic sugar for hooking things up to Messages system.
+    // We create separate subchannel strings for each operation within our general channelKey, instead of using
+    // a switch in the receiver.
+    var channelKey = "io.highfidelity.virtualBaton:" + batonName,
+        subchannelHandlers = {}, // Message channel string => {receiver, op}
+        subchannelKeys = {};     // operation => Message channel string
+    function subchannelKey(operation) { return channelKey + ':' + operation; }
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+        debugSend(destination, operation, data);
+        sendHelper(subchannelKey(operation) + destination, data);
+    }
+    function send(operation, data) {  // Send data for an operation on our channelKey.
+        debugSend('-', operation, data);
+        sendHelper(subchannelKeys[operation], data);
+    }
+    function messageHandler(channel, messageString, senderID) {
+        var handler = subchannelHandlers[channel];
+        if (!handler) { return; }
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 11]
```diff
@@ -0,0 +1,358 @@
...
+            }
+        } else {
+            send1('nack', data.proposerId, {proposalNumber: data.number});
+        }
+    });
+    // Paxos Learner behavior.
+    function localRelease() {
+        var callback = releaseCallback;
+        debugFlow('baton:', batonName, 'localRelease', 'callback:', !!releaseCallback);
+        if (!releaseCallback) { return; } // Already released, but we might still receive a stale message. That's ok.
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: examples/libraries/virtualBaton.js [hunk 12]
```diff
@@ -0,0 +1,358 @@
...
+        var holder = acceptedId();  // If we're waiting and we notice the holder is gone, ...
+        if (holder && claimCallback && !electionWatchdog && !connectionTest(holder)) {
+            bestPromise.winner = null; // used if the quorum agrees that old winner is not there
+            propose();  // ... propose an election.
+        }
+    }, recheckInterval);
+    exports.unload = function unload() { // Disconnect from everything.
+        messages.messageReceived.disconnect(messageHandler);
+        timers.clearInterval(exports.recheckWatchdog);
+        if (electionWatchdog) { timers.clearTimeout(electionWatchdog); }
```
Comment (ZappoMan) [inline examples/libraries/virtualBaton.js]: coding standard: braces on separate lines.

File: tests/mocha/test/testVirtualBaton.js [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+var virtualBaton = require('../../../examples/libraries/virtualBaton.js');
+
+describe('temp', function () {
+    var messageCount = 0, testStart = Date.now();
+    function makeMessager(nodes, me, mode) { // shim for High Fidelity Message system
+        function noopSend(channel, string, source) { }
+        function hasChannel(node, channel) { return -1 !== node.subscribed.indexOf(channel); }
+        function sendSync(channel, message, nodes, skip) {
+            nodes.forEach(function (node) {
+                if (!hasChannel(node, channel) || (node === skip)) { return; }
```
Comment (ZappoMan) [inline tests/mocha/test/testVirtualBaton.js]: coding standard: braces on separate lines.

File: tests/mocha/test/testVirtualBaton.js [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+            });
+        }
+        nodes.forEach(function (node) {
+            node.sender = node.sender || noopSend;
+            node.subscribed = node.subscribed || [];
+        });
+        return {
+            subscriberCount: function () {
+                var c = 0;
+                nodes.forEach(function (n) { if (n.subscribed.length) { c++; } });
```
Comment (ZappoMan) [inline tests/mocha/test/testVirtualBaton.js]: coding standard: braces on separate lines.

File: tests/mocha/test/testVirtualBaton.js [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+                return c;
+            },
+            subscribe: function (channel) {
+                me.subscribed.push(channel);
+            },
+            unsubscribe: function (channel) {
+                me.subscribed.splice(me.subscribed.indexOf(channel), 1);
+            },
+            sendMessage: function (channel, message) {
+                if ((mode === 'immediate2Me') && hasChannel(me, channel)) { me.sender(channel, message, me.name); }
```
Comment (ZappoMan) [inline tests/mocha/test/testVirtualBaton.js]: coding standard: braces on separate lines.

File: tests/mocha/test/testVirtualBaton.js [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+            this.timeout(5000);
+            var nodes = [{name: 'al'}, {name: 'bl'}],
+                gotA = false,
+                gotB = false,
+                releaseA = false;
+            makeBaton(testKeydsl, nodes, nodes[0], debug, mode, optimize).claim(function (key) {
+                console.log('claimed al', key);
+                assert.ok(!gotB, "should not get A after B");
+                gotA = true;
+                if (!gotB) { done(); }
```
Comment (ZappoMan) [inline tests/mocha/test/testVirtualBaton.js]: coding standard: braces on separate lines.

File: tests/mocha/test/testVirtualBaton.js [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+            }, function () {
+                assert.ok(gotA, "Should claim it first");
+                releaseA = true;
+                if (gotB) { done(); }
```
Comment (ZappoMan) [inline tests/mocha/test/testVirtualBaton.js]: coding standard: braces on separate lines.

File: tests/mocha/test/testVirtualBaton.js [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+            });
+            setTimeout(function () {
+                makeBaton(testKeydsl, nodes, nodes[1], debug, mode, optimize).claim(function (key) {
+                    console.log('claimed bl', key);
+                    gotB = true;
+                    if (releaseA) { done(); }
```
Comment (ZappoMan) [inline tests/mocha/test/testVirtualBaton.js]: coding standard: braces on separate lines.

**Comment by ZappoMan** [comment]
Other than the coding standard issues, I don't see anything glaring that would block the merge. I didn't test or study for correctness of the paxos implementation. But the use of messages seems reasonable.

**Comment by cozza13** [comment]
tested

**Comment by hifi-gustavo** [comment]
Build console output for commit: 31d72ffb8d86ba6b5f2971e3c121529d4b9a2fce available at: https://gist.github.com/db791c2c7496622bf07a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 989192e9fd1c4bd9d0875968b0e34408a4297d4c available at: https://gist.github.com/db791c2c7496622bf07a<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR7042.dmg

**Pull request merged**